### ISSUE TITLE
Feature (coroutines): Support yielding autoDisposing disposers (like Go defer)

### ISSUE THREAD

**Issue by spion** [petkaantonov/bluebird#1014]
Since we don't have using blocks in JavaScript, we don't currently have a good way to use disposers in generators.

Proposal: add support for yielding AutoDisposers:

```js
Promise.coroutine(function* () {
  var connection = yield getConnectionDisposer(options).autoDispose();
  // connection is usable here, and will dispose after coroutine completes
});
```

This would act like Go's defer:

```go
func DoSomething() err error {
    connection, err := db.Connect(options)
    if err != nil {
        return
    }
    defer connection.Close()
    // connection is usable here, and will close after function returns automatically.
}
```

Some thoughts:
* its nice for autoDispose to be explicit, otherwise its not clear that you can't save the connection elsewhere and use it later

**Comment by grabbou** [comment]
Re last point, would something like:
```js
Promise.coroutine(function* () {
  var connection = yield getConnectionDisposer(options).defer(conn => conn.close());
  // connection is usable here, and will dispose after coroutine completes
});
```
make sense?

**Comment by petkaantonov** [comment]
co.defer seems better but we need new name since defer might have something to do with deferreds

**Comment by spion** [comment]
@grabbou oh, nice and clean :) however I'm assuming `getConnectinDisposer` returns a bluebird disposer http://bluebirdjs.com/docs/api/disposer.html which already knows how the connection should be closed. 

We probably want to support existing disposers

**Comment by spion** [comment]
@petkaantonov Well, deferreds are deprecated anyway, right? And if its attached to `Promise.coroutine` its clearer that its a different sort of defer.

**Comment by grabbou** [comment]
Oh, you guys are quick! Actually decided to write on Twitter because I wasn't entirely sure about that idea yet.

Just for the reference, what I suggested was:
```js
var connection = yield getConnectionDisposer(options).defer(conn => conn.close());
```

**Comment by grabbou** [comment]
@petkaantonov another interesting idea would be to allow yielding deferred functions at any point inside a <s>go</s> coroutine by wrapping the function with co.defer

**Comment by petkaantonov** [comment]
Currently it will look like this for new code:

```js
import { coroutine as co } from "bluebird";

co(function copy* (srcPath, dstPath) {
    var src = yield fs.openAsync(srcPath);
    co.defer(_ => fs.closeAsync(src));
    var dst = yield fs.openAsync(dstPath);
    co.defer(_ => fs.closeAsync(dst));

    // File copy stuff that is not related but I wrote for fun
    const buffer = new Buffer(8092);
    let bytesRead, bytesWritten, totalBytesWritten;
    let totalBytesRead = 0;
    while (bytesRead = yield fs.readAsync(src, buffer, 0, buffer.length, totalBytesRead)) {
        bytesWritten = totalBytesWritten = 0;
        while (totalBytesWritten < bytesRead) {
            bytesWritten = yield fs.writeAsync(dst, buffer, totalBytesWritten,
                                                            bytesRead - totalBytesWritten,
                                                            totalBytesRead + totalBytesWritten);
            totalBytesWritten += bytesWritten;
        }

        totalBytesRead += bytesRead;
    }
});
```

And if youre already using disposers, they will have autoDisposer method that can be called inside coroutines:

```js
import { coroutine as co } from "bluebird";

function openFileAsync(path) {
    return fs.openAsync(path).disposer(function(fd) {
        return fs.closeAsync(fd);
    });
}

co(function copy* (srcPath, dstPath) {
    var src = yield openFileAsync(srcPath).autoDispose();
    var dst = yield openFileAsync(dstPath).autoDispose();
});
```

**Comment by petkaantonov** [comment]
Also: after coroutine completes (error, success, cancellation), the registered cleanup handlers are called in reverse order serially.

**Comment by benjamingr** [comment]
- What should happen if a `co.defer` throws?

On another topic - can we export `.coroutine` as `Promise.co` for convenience :P ?

**Comment by petkaantonov** [comment]
we should crash.. destructors must not throw

**Comment by vphantom** [comment]
Is this currently work in progress or a plan for the distant future?

I'm still early in a new project, trying to decide how I'll structure it. It looks like for now I'll be going for coroutines but like I used to do in C, I'll clean up resources in the context in which they were obtained:

```js
var doSomeDbStuff = Promise.coroutine(function *() {
  var conn = yield getFromPool();
  ...
  freeConnection(conn);
  return;
});
```

If I understand correctly what's being proposed here, with this addition my `getFromPool()` could return a Disposer for its Promise and Bluebird's coroutine handling would take care of calling the disposer method when the user coroutine itself returns? That sounds not only cleaner, but also safer.

```js
function getFromPool() {
  return (new Promise(function(resolve, reject) {
    // Acquire 'conn'
    return resolve(conn);
  })).disposer(function(conn) {
    conn.release();
  });
}

var doSomeDbStuff = Promise.coroutine(function *() {
  var conn = yield getFromPool();
  ...
  return;
});
```

If I understood this right, sounds like it might be worth waiting for. :)

**Comment by spion** [comment]
Yes thats what the feature is going to be :) There will also be `co.defer(_ => finallyBlockCode)`

Its just that I haven't had the time to finish the code. I'll try and finish it up in the next few days.

One thing that might be different is that the code will be:

```js
var doSomeDbStuff = Promise.coroutine(function *() {
  var conn = yield getFromPool().autoDispose();
  ...
  return;  // 'conn' would be released around here
});
```

Otherwise, its not quite clear that you can't keep the reference to the connection after the coroutine completes.

**Comment by vphantom** [comment]
Very nice. :) Meanwhile I learned how to use disposers in an overall coroutine-centric design based on the `withTransaction(fn)` example. (Coroutines being themselves promises was the key.) It'll be easy to refactor from using those wrappers into yield `xxx().autoDispose()` in due time, and allows me to use yields for everything not needing disposal.

**Comment by benjamingr** [comment]
@spion still pursuing this?

**Comment by spion** [comment]
Yes. I'll try to finish it this week. Sorry for the long delay.

**Comment by overlookmotel** [comment]
By the way, concerning @benjamingr's suggestion:

> On another topic - can we export .coroutine as Promise.co for convenience :P ?

`Promise.coroutine()` is same as [co](https://www.npmjs.com/package/co)'s `co.wrap()`. With [co](https://www.npmjs.com/package/co), `co()` creates a coroutine and *executes it immediately*; `Promise.coroutine()` just returns a coroutine function ready to be called.
